CNDB-15609 test SAI disk size for all versions

k-rus opened the pull request:
### What is the issue
Switching to row aware formats significantly increased size of SAI files. This wasn't expected and created incidents.

### What does this PR fix and why was it fixed

Fixes https://github.com/riptano/cndb/issues/15609

Adds tests to control simple SAI disk size for different format version and different partition sizes, when data are split into 2 SSTables and after compaction into a single SSTable.
It uses expected size as upper bound and has 5% less for lower bound. The upper bounds were found empirically. The goal of the test is too fail if disk size increases or significantly decreases.

This also should help to see improvements in [#15608 ](https://github.com/riptano/cndb/issues/15608).

Vector and analyzed SAI is outside the scope of this PR.

k-rus committed:
CNDB-15609 test SAI disk size for all versions

Add tests to control SAI disk size for different format version and
different partition sizes.

k-rus committed:
CNDB-15609 assert lower bound on SAI disk size

It will prevent reporting incorrect value such as 0 and also notice
improvements.

k-rus committed:
Fix bug in populating segments and polish

k-rus commented:
> * Make sure there is a PR in the CNDB project updating the Converged Cassandra version

This PR adds a test and doesn't modify non-test code, thus no CNDB PR is created.

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,186 @@
+                                      return Stream.of(
+                                      new Object[]{ v, 24991, "pk", 1 },
+                                      new Object[]{ v, 26183, "pk, v_int", 2 },
+                                      new Object[]{ v, 28530, "pk, v_int", 100 });
```
Can you add a comment where those expectedDiskSize numbers are coming from?
Are those just the empirical index sizes you measured in this test? 
Also, why do we expect those tiny differences e.g. between d and e and f versions?

k-rus replied:
I added a comment that the sizes were obtained empirically and refer to Version to understand the difference. I also added logging to log actual sizes.

> why do we expect those tiny differences e.g. between d and e and f versions?
There are changes in the components stored on disk. Changes to Vector disk format do not affect the test as no vectors are included (see https://github.com/riptano/cndb/issues/17471). I also added this briefly to the comment.

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,186 @@
+        .isGreaterThan((long) (expectedDiskSize * 0.8));
+    }
+
+    private void insertRows(int size, int start) throws UnknownHostException
```
`size` doesn't tell me much how many rows will be inserted. Is it the number of rows? Or partitions? Or something else?

From the code I conclude those are just rows. Please rename to `rowCount` for better clarity.
Similarly `start` looks like the first partition key. Maybe `firstPartitionKey`?
Or no, actually I think you meant `start` to be the `firstRow`.

k-rus replied:
Renamed to `nrRows` and `startRow`. `firstRow` is not intuitive for me.

k-rus replied:
Let me know if new names are suitable.

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,186 @@
+                    "v_decimal, v_double, v_float, v_text, v_timestamp, v_uuid, v_varchar, " +
+                    "v_varint, v_timeuuid, v_inet, v_date, v_time, v_smallint, v_tinyint, v_duration) " +
+                    "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
+                    start + i % (size / rowsPerPartition),
```
mathematical nit: if `size < rowsPerPartition` this is going to divide by 0.
Maybe you wanted to round the division up? 

`partitionCount = Math.round(Math.ceil(size / rowsPerPartition))`

And one more thing:
why add `start` here, if `i` already starts at `start` ?

k-rus replied:
I don't want any roundings here. It's a test. I added assertions that the number of rows is dividable by rows per partition and the result is non-zero.

pkolaczk replied:
Well, you already had rounding. The code was implicitly rounding down so would end up by off-by-one number of partitions if the rows/rows-per-partition were not divisible exactly. That's why in the edge case it would hit divide-by-zero (where divide by 1 would make a lot more sense to me). If you insert 10 rows and specify 100 rows per partition, you end up with 1 partition of 10 rows. Partition count = round_up(rows / rows-per-partition). 

In a test I'm expecting I can change parameters freely, so I'm not expecting the test to artificially crash if I set number of rows to 0, to 1 or to a number not divisible by sth. As long as the number of rows makes sense (i.e. is not negative or crazy high), it should IMHO work. Therefore I'm not very happy with the assert either. It just adds unnecessary friction when working with such test and experimenting with the inputs.

k-rus replied:
> In a test I'm expecting I can change parameters freely, so I'm not expecting the test to artificially crash if I set number of rows to 0, to 1 or to a number not divisible by sth. As long as the number of rows makes sense (i.e. is not negative or crazy high), it should IMHO work. Therefore I'm not very happy with the assert either. It just adds unnecessary friction when working with such test and experimenting with the inputs.

I suggest adjusting for this need is done when needed. The test is not written in stone and can be extended to cover more use cases :)

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,186 @@
+                    (float) (i * 2.5),
+                    "text_value_" + i,
+                    new Date(System.currentTimeMillis() + i * 1000L),
+                    UUID.randomUUID(),
```
Consider using a fixed random seed so the test is repeatable.

k-rus replied:
It doesn't affect the test. UUID is fixed size and is not indexed or used in queries.

pkolaczk replied:
But now you're relying on a particular property of the code under test. 
Yes, UUIDs are fixed size. But if we we ever add compression, their content will affect the size.
I'd prefer to have test repeatable in the most direct way - by generating the same data for the same seed, not relying on some secondary effects that some code 10 layers down is not sensitive to the input values. Having repeatable data also makes debugging easier (I can record one execution and compare with another; but if inputs are changing this adds noise).

k-rus replied:
Using fixed seed requires additional verbosity to my understanding. IMHO it's not worth the effort.

k-rus replied:
Can randomness of this values affect compression results between runs? Does compression perform across rows?

pkolaczk replied:
If you generate different data in each run, a compression can compress them to a different size.

> Does compression perform across rows?

Yes, obviously.
Compression is by blocks, which are not aligned to row/partition boundaries.

pkolaczk replied:
> Using fixed seed requires additional verbosity to my understanding. IMHO it's not worth the effort.

You mean initializing a single Random field is a lot verbosity?
Or using the class that was already supposed to solve that problem - `SaiRandomizedTest`?
If `SaiRandomizedTest` is not useful for your purpose here, then maybe think how to improve it, because it is useful for the other tests. We don't want each test to reinvent their own way of generating random data ;)

k-rus replied:
> You mean initializing a single Random field is a lot verbosity?

Bob produced a lot of code when I asked him for this.

> Or using the class that was already supposed to solve that problem - `SaiRandomizedTest`?
> If `SaiRandomizedTest` is not useful for your purpose here, then maybe think how to improve it, because it is useful for the other tests. We don't want each test to reinvent their own way of generating random data ;)

I can investigate this option. This test is not purely randomized, so it can be confusing.

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,186 @@
+        assertThat(diskSize)
+        .as("Disk size for SAI version %s before compaction", version)
+        .isLessThanOrEqualTo(expectedDiskSize)
+        .isGreaterThan((long) (expectedDiskSize * 0.8));
```
`expectedDiskSize` values have suspiciously many significant digits in the input data for the test (looks like they were given with precision to 1 byte). This looks to me a bit fragile. 

If the goal of this test is to catch any change in the size of the data, why simply not check for exact equality and get rid of randomness in the test data? That's not necessarily a bad idea, as it would catch any changes in how indexes are written. 

But if the goal is to test if index size is *reasonable* then why not something more symmetric allowing a bit of variability in both directions?
```
   .isLessThanOrEqualTo(expectedDiskSize * 1.05)
   .isGreaterThan((long) (expectedDiskSize * 0.95));
```

Im afraid that if you're comparing the bytes exactly with the precision of 1 byte, then if we ever add a new version of the index which is going to be 0.0001% larger, this test will need changes. 

On the other hand, the 0.8 on the lower bound feels a bit too relaxed. If we significantly reduce the index size in the next release e.g. to 0.85 of the current release, then the test won't catch a regression of going back to 0.99. Also a big unexpected change in the index size would be suspicious, even if the index is much smaller.

k-rus replied:
> get rid of randomness in the test data

There is no randomness in the data size. UUID is fixed in size. I don't see other cases for randomness.

k-rus replied:
I understand the concern about fragility of the test. So far the reason for the range is that flush and compaction have minor differences in the size.
It's easier to specify an upper bound.

> Im afraid that if you're comparing the bytes exactly with the precision of 1 byte, then if we ever add a new version of the index which is going to be 0.0001% larger, this test will need changes.

The purpose of the test is to catch increases in the size, so correct the test will need to be updated and cover the new format. Also I want to catch significant reductions, like in https://github.com/riptano/cndb/issues/15609. This is intentional.

pkolaczk replied:
All I'm saying this test won't catch increases in size if over time we reduce the size. If the size gets reduced by 20% then you don't have to adapt this test and at the same time it loses its ability to catch increase in size.

k-rus replied:
> All I'm saying this test won't catch increases in size if over time we reduce the size. If the size gets reduced by 20% then you don't have to adapt this test and at the same time it loses its ability to catch increase in size.

This is a good point.

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,186 @@
+    @Parameterized.Parameter(3)
+    public int rowsPerPartition;
+
+    @Parameterized.Parameters(name = "saiFormat={0}, expectedDiskSize={1}, pkColumns={2}, partitionSize={3}")
```
nit: can you use the same name for the same thing? Either `rowsPerPartition` or `partitionSize` but not both in the same code. And... I like `rowsPerPartition` better, as it's more specific.

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,186 @@
+    {
+        return Version.ALL.stream()
+                          .flatMap(v -> {
+                              switch (v.toString())
```
Suggestion: We don't expect on-disk index format changes between minor versions, so maybe it's enough to switch on the first letter only? That would simplify the switch body.

k-rus replied:
> Suggestion: We don't expect on-disk index format changes between minor versions, so maybe it's enough to switch on the first letter only? That would simplify the switch body.

We do have changes between minor format changes. To my understanding it's allowed in the case of adding additional information. See move from `ec` to `ed`.

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,186 @@
+                    i * 1.5,
+                    (float) (i * 2.5),
+                    "text_value_" + i,
+                    new Date(System.currentTimeMillis() + i * 1000L),
```
Maybe use a fixed time constant here instead of current time?

pkolaczk reviewed (changes requested):
Overall I like the idea of adding this test.

The biggest issue is that it has some random elements and at the same time seems to compare the sizes very strictly on the upper bound. If we ever add a tiny amount of data to the indexes, or if the index size becomes a function of the data values in the future (like it would be if we ever introduce some compression), this test will likely  break.

k-rus committed:
Improve understability of the test

Improve names of methods and variables. Added comments. Added asserts.
Added logging of actual sizes.

k-rus committed:
Rename variables to improve readability

k-rus committed:
Use constant time instead of today

k-rus commented:
> The biggest issue is that it has some random elements

Can you explain how the random UUID can affect the disk size?



> at the same time seems to compare the sizes very strictly on the upper bound. If we ever add a tiny amount of data to the indexes, (like it would be if we ever introduce some compression), this test will likely break.

And the goal of the test is to break in this case. Any new format or changes in existing format, which go over upper limit or significantly reduce the size, should blow up the test.

>if the index size becomes a function of the data values in the future (like it would be if we ever introduce some compression), this test will likely break.

When this will happen, we will need to reimplement the test. IMHO it's not worth the effort to count for this now.

k-rus commented:
I made slight improvement to PR description.

pkolaczk reviewed (approved):
Considering this is just a test and it works, I get the point we don't want to spend too much time on this now.

k-rus committed:
Update name of expected disk size

k-rus committed:
Use IBM copyright

k-rus committed:
Separate expected sizes before and after compaction

It allows to reduce the expected size interval.

k-rus commented:
@pkolaczk I separated expected sizes before and after compaction. It allowed to reduce the expected interval. See the last commit, [314d9f3](https://github.com/datastax/cassandra/pull/2045/commits/314d9f316b914963faa411ba38e6b93da9b26a2a).
I see some fluctuations in sizes for AA between runs. Also the difference in disk sizes between before and after compaction might rise new questions. Do you think there should be followed up issues and which? I prefer to merge this PR and have separate follow up issues.

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/index/sai/functional/SaiDiskSizeTest.java:
```diff
@@ -0,0 +1,209 @@
+                                     "expectedDiskSizeForCompactedSSTable={2}, pkColumns={3}, rowsPerPartition={4}")
+    public static Collection<Object[]> generateParameters()
+    {
+        return Version.ALL.stream()
```
@k-rus, I have a suggestion for this test. When people add new index versions, not everyone may be aware or remember about this test. Should we add an assertion on the versions in Version, so that if someone adds a new one, the test fails and they are reminded they need to update it for a new version? I can open a ticket if you agree

k-rus replied:
@ekaterinadimitrova2 sorry for delay in my reply. I don't have a strong opinion on this. I guess the main case to prevent adding new SAI format without updating the test is when new format slightly improve or doesn't change the disk utilization. Is it important to catch all new SAI format cases? In such ase it can be done by throwing or asserting in the default case.
